### ISSUE TITLE
DNS confused when switching WiFi networks

### ISSUE THREAD

**Issue by mcandre** [boot2docker/boot2docker#776]
When I switch WiFi networks, `docker pull` times out, as well as `apt-get update` inside containers.

As a mitigation, I can manually `boot2docker restart`. In the future, can we improve networking so that Docker containers gracefully change with their hosts, for more mobile users?
- Docker 1.5
- boot2docker 1.5
- Mac OS X 10.10.2 Yosemite

docker/docker/issues/11220

**Comment by cybercypher** [comment]
I have the same issue

**Comment by egoldschmidt** [comment]
+1, same here

**Comment by bnvk** [comment]
Yep, I just experienced this same issue when switching from home to work Wifi networks. Our devops wizard had to help me resolve the issue. The `boot2docker restart` resolved the issue- but for a non sys admin front-end engineer like me I was totally stumped!

**Comment by ehajri** [comment]
I have just spent 4 hours working on why my containers could not connect to the internet. Nothing worked.
Then --dns=8.8.8.8 fixed my problem and with [further reading](https://docs.docker.com/articles/networking/#configuring-dns), I found out my container's /etc/resolve.conf contains my **home wifi** gateway, while now I am at the office using a different connect.

A simple `boot2docker down` and `boot2docker up` fixed it.

**Comment by tnolet** [comment]
This is indeed a major issue. I have to restart boot2docker constantly just to "reset" the DNS resolving.

**Comment by apobbati** [comment]
Yup. This is becoming a constant problem and when I have running containers, it makes life difficult.

**Comment by randerzander** [comment]
Has anyone figured out a workaround short of manually editing /etc/resolve.conf inside every running container?

**Comment by rburgst** [comment]
Not sure if this helps but I am using https://github.com/Parallels/boot2docker-vagrant-box and there the problem does not appear. Whether this is a problem in VBox or boot2docker is hard for me to say.

**Comment by egoldschmidt** [comment]
Setting my Mac's default DNS entries to Google's public DNS servers
(8.8.8.8, 8.8.4.4) solved the issue. It's a bit heavy handed but it works.

On Friday, August 7, 2015, rburgst <notifications@github.com> wrote:

> Not sure if this helps but I am using
> https://github.com/Parallels/boot2docker-vagrant-box and there the
> problem does not appear. Whether this is a problem in VBox or boot2docker
> is hard for me to say.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/boot2docker/boot2docker/issues/776#issuecomment-128912639>
> .
>


-- 
(phone)

**Comment by lukemarsden** [comment]
This is still a problem, any chance for a fix? It seriously impinges on Docker's usability for a large proportion of users, IMO.

**Comment by isubuz** [comment]
Even `boot2docker restart` isn't fixing the issue for me. This is really a blocker! Are there any plans to fix this?

**Comment by tmajibon** [comment]
For those encountering this where it's not a DNS issue:

Check your iptables rules after a docker restart and immediately after a network disconnect/reconnect.

If you have your iptables rules getting restored upon reconnect and the configuration saved was prior to the docker install, your system might be removing the docker NAT rules after every reboot.

I'm just sharing this as this was the problem I just solved on my own install and it's really easy to overlook. 

I had a script in /etc/network/if-up.d/ restoring my iptables from a file on every reboot. That file had not been updated since installing docker, so every disconnect resulted in that script running and knocking out the docker nat ruleset. (these disconnects including sleeping the computer, or just directly switching networks)

**Comment by tmajibon** [comment]
An alternative to consider if all networking fails:

Check to see if iptables is being restored on network connection being established.

In my own system I had a script in /etc/network/if-up.d/ restoring the iptables on network connection.

The restored rules did not include Docker's nat rules, which broke all docker network access every time the system reconnected.

By saving a fresh copy of iptables after restarting docker, the problem was solved for me.

Sharing this here as I found this while looking for answers and it's a really easy mistake to make.
